packages/stateManager/index.ts

```diff
@@ -36,67 +101,345 @@ abstract class PublicObjectManager<
   // returns a bool to indicate whether the event was processed or not
   async _processEvent(event: schema.ShopEvents): Promise<boolean> {
-    if (typeof event === schema.ShopEvents.UpdateItem) {
-      // load the item from the store
-      const item = await this.db.get(event.eventId);
-      Object.assign(item, event);
-      await this.db.put(event.eventId, item);
-      this.emit("update", item);
+    if (event.createItem) {
+      const ci = event.createItem;
+      const metadata = JSON.parse(new TextDecoder().decode(ci.metadata));
+      const id = bytesToHex(ci.eventId);
+      const item = {
+        price: ci.price,
+        metadata,
+        tags: [],
+        quantity: 0,
+      };
+      await this.db.put(id, item);
+      return true;
+    } else if (event.updateItem) {
+      const ui = event.updateItem;
+      const id = bytesToHex(ui.itemId);
+      const item = await this.db.get(id);
+      if (ui.metadata) {
+        item.metadata = JSON.parse(new TextDecoder().decode(ui.metadata));
+      }
+      if (ui.price) {
+        item.price = ui.price;
+      }
+      await this.db.put(id, item);
+      return true;
+    } else if (event.changeStock) {
+      const cs = event.changeStock;
+      if (cs.itemIds) {
+        cs.itemIds.map(async (id: Uint8Array, i: number) => {
+          const itemId = bytesToHex(id);
+          const item = await this.db.get(itemId);
+          const diff = cs.diffs ? cs.diffs[i] : 0;
+          item.quantity = item.quantity + diff;
+          await this.db.put(itemId, item);
+        });
+        // if the changeStock event is emitted from a payment(AKA has txHash), it has to also go through OrderStore
+        if (cs.txHash) {
+          return false;
+        } else return true;
+      }
+    } else if (event.updateTag) {
+      const ut = event.updateTag;
+      const tagId = bytesToHex(ut.tagId);
+
+      if (ut.addItemId) {
+        const itemId = bytesToHex(ut.addItemId);
+        const item = await this.db.get(itemId);
+        item.tags = [...item.tags, tagId];
+        await this.db.put(itemId, item);
+      }
+      if (ut.removeItemId) {
+        const itemId = bytesToHex(ut.removeItemId);
+        const item = await this.db.get(itemId);
+        item.tags = [...item.tags.filter((id: `0x${string}`) => id !== tagId)];
+        await this.db.put(itemId, item);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  async create(item: Item) {
+    const id = bytesToHex(
+      await this.client.createItem({
+        price: item.price,
+        metadata: new TextEncoder().encode(JSON.stringify(item.metadata)),
+      }),
+    );
+    await this.db.put(id, item);
+  }
+
+  async update(
+    itemId: `0x${string}`,
+    price?: string,
+    metadata?: { name: string; description: string; image: string },
+    quantity?: number,
+  ) {
+    const item = await this.db.get(itemId);
```

Comment:
do this in parallel with network operation so that when the relay has confirmed the update we will have the data from the db.